### ISSUE TITLE
`X509_dup()` should better copy the cached (non-ASN.1) data

### ISSUE THREAD

**Issue by DDvO** [openssl/openssl#23337]
This is a follow-up on https://github.com/openssl/openssl/pull/23043:

@vdukhovni: (https://github.com/openssl/openssl/pull/23043#discussion_r1457745827)
I see this was already merged, but I am curious why it is that absence of lazily-computed data causes observable behaviour differences. If that's the case, perhaps the real issue is not in documentation, but API semantics.

When the computed data is missing, the relevant thread should lock the object and recompute the missing data. Or else, the data should automatically be computed as part of X509_dup().

Why should users need to know about these sort of rather subtle behaviour differences?

@DDvO: (https://github.com/openssl/openssl/pull/23043#discussion_r1457787283)
Indeed to me it was a pretty intricate API pitfall that `X509_dup()` does not copy everything.
https://github.com/siemens/libsecutils/pull/57 contains details on where it showed up.

**Comment by DDvO** [comment]
I fear that the doc change done in  #23043 was still needed at least for the older OpenSSL versions 
(where we presumably won't touch the semantics of `X509_dup()`).

**Comment by t8m** [comment]
Yeah, we need to be careful here in terms of backwards compatibility. I would call this a feature and thus #23043 was necessary anyway.
